src/components/ha-hls-player.ts

```diff
@@ -126,13 +125,19 @@ class HaHLSPlayer extends LitElement {
     }
 
     this._useExoPlayer = await useExoPlayerPromise;
-    let hevcRegexp: RegExp;
-    let masterPlaylist: string;
-    if (this._useExoPlayer) {
-      hevcRegexp = /CODECS=".*?((hev1)|(hvc1))\..*?"/;
-      masterPlaylist = await (await masterPlaylistPromise).text();
+    const masterPlaylist = await (await masterPlaylistPromise).text();
+
+    // use regular playlist instead of master playlist if possible
+    let playlist_url: string;
+    const playlistRegexp = /#EXT-X-STREAM-INF:.*?(?:CODECS=".*?(hev1|hvc1)?\..*?".*?)?(?:\n|\r\n)(.+)/g;
+    const match = playlistRegexp.exec(masterPlaylist);
+    if (match !== null && playlistRegexp.exec(masterPlaylist) === null) {
+      playlist_url = new URL(match[2], this.url).href;
```

Comment:
Please explain to me, and in a comment in the code, what we are doing here.

Reply:
Sure, I added comments to the code which should help clarify things.
HLS has two types of playlists, a master playlist and regular playlists. The master playlist usually contains multiple variant streams and metadata which describes them (including bandwidth and resolution). The players can then choose between the variants depending on available resources.
The stream component originally just used a single regular playlist since we only have one stream variant. In 0.115, we made several changes to stream, including not including an audio stream with the video stream when there is no audio. However, this made chromecast devices not play the stream. We realized that we could fix around this problem by making a master playlist with only our single variant regular playlist with metadata which indicated that the stream had only video and no audio.
After this change, the stream component now sends a master playlist instead of a regular playlist. This is slightly less efficient since a master playlist must be loaded and parsed before the regular playlist can be loaded and played. Using a master playlist did help us by giving the ability to check the codec of the stream so we could restrict ExoPlayer use to only videos using the H.265 (HEVC) codec. But this actually resulted in 2 loads of the master playlist before the regular playlist load - once by ha-hls-player to check the codec, then once by the player itself. So we effectively went from one load (the player loading the regular playlist) to three loads (ha-hls-player loading the master playlist, the player loading the master playlist, then the player loading the regular playlist).
We previously reduced these 3 loads to 2 loads by passing the regular playlist directly to the player, helping the player avoid loading the master playlist. However, we did this by a simple string replace. @hunterjm noted that this might be undesirable because makes the frontend rely on the path of the variant stream coded in the backend.
This PR just gets the regular playlist correctly by parsing the regular playlist url from the master playlist.